internal/orgcollector/orgcollector.go

```diff
@@ -0,0 +1,147 @@
+// NOTE: whenever the CC API V3 exposes the quota information for us,
+// we should just store everything in the app cache and get it from there
+package orgcollector
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-firehose-nozzle/internal/client/cloudfoundry"
+	"github.com/DataDog/datadog-firehose-nozzle/internal/config"
+	"github.com/DataDog/datadog-firehose-nozzle/internal/metric"
+	"github.com/DataDog/datadog-firehose-nozzle/internal/util"
+
+	"github.com/cloudfoundry-community/go-cfclient"
+	"github.com/cloudfoundry/gosteno"
+)
+
+type OrgCollector struct {
+	cfClient              *cloudfoundry.CFClient
+	log                   *gosteno.Logger
+	processedMetrics      chan<- []metric.MetricPackage
+	customTags            []string
+	queryInterval		  uint32
+	stopper               chan bool
+}
+
+func NewOrgCollector(
+	config *config.Config,
+	processedMetrics chan<- []metric.MetricPackage,
+	log *gosteno.Logger,
+	customTags []string) (*OrgCollector, error) {
+	cfClient, err := cloudfoundry.NewClient(config, log)
+	if err != nil {
+		return nil, err
+	}
+	return &OrgCollector{
+		cfClient:         cfClient,
+		log:              log,
+		processedMetrics: processedMetrics,
+		customTags:       customTags,
+		queryInterval:    config.OrgDataQuerySeconds,
+		stopper:          make(chan bool),
+	}, nil
+}
+
+func (o *OrgCollector) Start() {
+	go o.run()
+}
+
+func (o *OrgCollector) Stop() {
+	o.stopper <- true
+}
+
+func (o *OrgCollector) run() {
+	ticker := time.NewTicker(time.Duration(o.queryInterval) * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			go o.pushMetrics()
+		case <-o.stopper:
+			return
+		}
+	}
+}
+
+func (o *OrgCollector) pushMetrics() {
+	var wg sync.WaitGroup
+	errors := make(chan error, 10)
+	// Fetch orgs
+	wg.Add(1)
+	var allOrgs []cfclient.Org
+	go func() {
+		defer wg.Done()
+		var err error
+		allOrgs, err = o.cfClient.GetV2Orgs()
+		if err != nil {
+			errors <- err
+		}
+	}()
```

Comment:
Can't we use the org information that is already cached, instead of fetching it every time ?

Reply:
I wanted to do that, but IMO it's not worth the effort right now. Let me explain: We either get app cache from V3 entirely or V2 entirely.

* If V3 is available (which is true most often on currently supported CF versions), we get everything from V3
  * V3 only has experimental support for returning quota GUID as a member of the org structure and it's not even present in all supported CF versions, so we can't rely on that anyway. Quota definitions are not available through V3 in *any* CF version yet.
  * This means V3 is out of question.
* If only V2 is available, we query only the app endpoint and make it return related values (including orgs). So we'd have to special case operations with V2 to also store the Orgs, which wouldn't be completely straightforward and it would bring little value (IMO), since we all currently supported CF versions should already have V3 and we can't use that for orgs as explained above.

Basically I think we should do this at the point when all supported CF versions have quotas available through the V3 API (which will probably take a while) - I also made a comment about this at the top of the `orgcollector.go`.

Reply:
OK fair enough. Thanks for the explanation. Do we have an idea of the cardinality of orgs in the bigger infrastructures ?

Reply:
From what I've been told it's in high hundreds, maybe a thousand or a little bit more.